### DISCUSSION TITLE
ERROR: 02_pytorch_classification --> error in 8.5 Training and Testing for multi-class Pytorch model

### DISCUSSION THREAD

**Discussion by fivefishstudios** [mrdbourke/pytorch-deep-learning#146] [Q&A] [answered]
In section 8.5, (multi-class classification model), this section produces an error message 

`# Fit the model
torch.manual_seed(42)

# Set number of epochs
epochs = 100

# Put data to target device
X_blob_train, y_blob_train = X_blob_train.to(device), y_blob_train.to(device)
X_blob_test, y_blob_test = X_blob_test.to(device), y_blob_test.to(device)

for epoch in range(epochs):
    ### Training
    model_4.train()

    # 1. Forward pass
    y_logits = model_4(X_blob_train) # model outputs raw logits 
    y_pred = torch.softmax(y_logits, dim=1).argmax(dim=1) # go from logits -> prediction probabilities -> prediction labels
  
    # 2. Calculate loss and accuracy
    loss = loss_fn(y_pred, y_blob_train) 
    acc = accuracy_fn(y_true=y_blob_train,
                      y_pred=y_pred)

    # 3. Optimizer zero grad
    optimizer.zero_grad()

    # 4. Loss backwards
    loss.backward()

    # 5. Optimizer step
    optimizer.step()

    ### Testing
    model_4.eval()
    with torch.inference_mode():
      # 1. Forward pass
      test_logits = model_4(X_blob_test)
      test_pred = torch.softmax(test_logits, dim=1).argmax(dim=1)
      # 2. Calculate test loss and accuracy
      test_loss = loss_fn(test_logits, y_blob_test)
      test_acc = accuracy_fn(y_true=y_blob_test,
                             y_pred=test_pred)

    # Print out what's happening
    if epoch % 10 == 0:
        print(f"Epoch: {epoch} | Loss: {loss:.5f}, Acc: {acc:.2f}% | Test Loss: {test_loss:.5f}, Test Acc: {test_acc:.2f}%") 
`

The error is: Expected floating point type for target with class probabilities, got Long

I'm on OSX, Pytorch '1.12.1' 

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-57-322e052845d1> in <module>
     18 
     19     # 2. Calculate loss and accuracy
---> 20     loss = loss_fn(y_pred, y_blob_train)
     21     acc = accuracy_fn(y_true=y_blob_train,
     22                       y_pred=y_pred)

~/opt/anaconda3/envs/MachineLearning/lib/python3.8/site-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
   1128         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130             return forward_call(*input, **kwargs)
   1131         # Do not call functions when jit is used
   1132         full_backward_hooks, non_full_backward_hooks = [], []

~/opt/anaconda3/envs/MachineLearning/lib/python3.8/site-packages/torch/nn/modules/loss.py in forward(self, input, target)
   1162 
   1163     def forward(self, input: Tensor, target: Tensor) -> Tensor:
-> 1164         return F.cross_entropy(input, target, weight=self.weight,
   1165                                ignore_index=self.ignore_index, reduction=self.reduction,
   1166                                label_smoothing=self.label_smoothing)

~/opt/anaconda3/envs/MachineLearning/lib/python3.8/site-packages/torch/nn/functional.py in cross_entropy(input, target, weight, size_average, ignore_index, reduce, reduction, label_smoothing)
   3012     if size_average is not None or reduce is not None:
   3013         reduction = _Reduction.legacy_get_string(size_average, reduce)
-> 3014     return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
   3015 
   3016 

RuntimeError: Expected floating point type for target with class probabilities, got Long

**Comment by tonyhaya** [comment]
The loss don't receive y_pred, instead use y_logit

**Reply by Arsive02** [reply]
"""The combination of nn.LogSoftmax and nn.NLLLoss is equivalent to using nn.CrossEntropyLoss. This terminology is a particularity of PyTorch, as the nn.NLLoss [sic] computes, in fact, the cross entropy but with log probability predictions as inputs where nn.CrossEntropyLoss takes scores (sometimes called logits). Technically, nn.NLLLoss is the cross entropy between the Dirac distribution, putting all mass on the target, and the predicted distribution given by the log probability inputs."""
Reason

**Comment by tonyhaya** [comment, marked as answer]
`loss = loss_fn(y_logits, y_blob_train) `
` acc = accuracy_fn(y_true=y_blob_train,
                  y_pred=y_pred)`

**Reply by mrdbourke** [reply]
Thank you for helping!
